File: apps/apple/Sources/Features/Admin/UsersPage.swift
import SwiftUI

/// Native SwiftUI parity of `web/src/features/admin/pages/UsersPage.tsx` — the admin "Subjects"
/// impersonation page. Reproduces the web `PageContainer` chrome (title + subtitle) wrapping the
/// single web `GlassPanel` (GlassPanel1), whose body renders one of the five branches the web page
/// switches on:
///  - open-mode note (web `open ? <div>…openMode…</div>`);
///  - loading (web `candidates.isLoading ? <Spinner />`);
///  - error + retry (web `candidates.isError ? <ErrorDisplay onRetry … />`);
///  - empty (web `subjects.length === 0 ? <EmptyState … />`);
///  - the subjects list, each row pairing the opaque subject with the embedded
///    `UserImpersonateButton` (web `<UserImpersonateButton subject disabled={active} />`).
///
/// Adaptive (ADR-002/006): a single leading-aligned column inside a `ScrollView` that fills the
/// regular-width macOS/iPad canvas and the compact iPhone width alike; each subject row stacks its
/// monospaced identifier above its action block so the richer native button never fights the row.
/// Every visible literal resolves from `Localizable.xcstrings` with the web key names verbatim
/// (`impersonation.users.*`); data binds through the `@Observable` `UsersPageModel` — no networking
/// in the view. There are no units on this surface (web `numberFormat`/SI converters unused).
public struct UsersPage: View {
    @State private var model: UsersPageModel

    public init(model: UsersPageModel) {
        _model = State(initialValue: model)
    }

    public var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: TSSpacing.x2xl) {
                header
                TSGlassPanel { panelBody }
                    .accessibilityElement(children: .contain)
                    .accessibilityLabel(Text("translation.impersonation.users.title"))
            }
            .padding(TSSpacing.lg)
            .frame(maxWidth: .infinity, alignment: .leading)
        }
        .background(Color.TS.bg.ignoresSafeArea())
        .task { model.load() }
    }

    // MARK: - Header (web PageContainer title + subtitle)

    private var header: some View {
        VStack(alignment: .leading, spacing: TSSpacing.xs) {
            TSPageTitle("translation.impersonation.users.title")
            Text("translation.impersonation.users.subtitle")
                .font(Font.TS.body)
                .foregroundStyle(Color.TS.textSecondary)
                .fixedSize(horizontal: false, vertical: true)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .accessibilityElement(children: .combine)
    }

    // MARK: - Panel body (web GlassPanel branch)

    @ViewBuilder
    private var panelBody: some View {
        switch model.panelState {
        case .openMode:
            openModeNote
        case .loading:
            loadingState
        case let .error(message):
            errorState(message)
        case .empty:
            emptyState
        case let .loaded(subjects):
            subjectsList(subjects)
        }
    }

    /// Web open-mode branch: a secondary-text note explaining impersonation needs forward-auth.
    private var openModeNote: some View {
        Text("translation.impersonation.users.openMode")
            .font(Font.TS.body)
            .foregroundStyle(Color.TS.textSecondary)
            .fixedSize(horizontal: false, vertical: true)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.vertical, TSSpacing.sm)
            .accessibilityIdentifier("users-page-open-mode")
    }

    /// Web candidates `isLoading`: a centered spinner.
    private var loadingState: some View {
        HStack {
            Spacer(minLength: 0)
            TSSpinner(label: "translation.impersonation.users.title")
            Spacer(minLength: 0)
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, TSSpacing.x2xl)
        .accessibilityIdentifier("users-page-loading")
    }

    /// Web candidates `isError`: the shared error display with a retry that refetches candidates.
    private func errorState(_ message: String) -> some View {
        TSErrorDisplay(onRetry: { model.retryCandidates() })
            .frame(maxWidth: .infinity)
            .accessibilityValue(Text(verbatim: message))
            .accessibilityIdentifier("users-page-error")
    }

    /// Web `subjects.length === 0`: the big empty state (no remediation — wait for another sign-in).
    private var emptyState: some View {
        TSEmptyState(
            title: "translation.impersonation.users.emptyTitle",
            message: "translation.impersonation.users.emptyMessage",
            systemImage: "person.2.slash"
        )
        .frame(maxWidth: .infinity)
        .accessibilityIdentifier("users-page-empty")
    }

    // MARK: - Subjects list (web `<ul className="divide-y">` of rows)

    private func subjectsList(_ subjects: [ImpersonationSubject]) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            ForEach(Array(subjects.enumerated()), id: \.element.id) { index, subject in
                if index > 0 { Divider() }
                subjectRow(subject)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .accessibilityIdentifier("users-page-list")
    }

    /// One subject row: the monospaced opaque identifier (web `font-mono break-all`) above its
    /// embedded `UserImpersonateButton`, which owns its own gated action + every state.
    private func subjectRow(_ subject: ImpersonationSubject) -> some View {
        VStack(alignment: .leading, spacing: TSSpacing.sm) {
            Text(verbatim: subject.subject)
                .font(.system(.subheadline, design: .monospaced))
                .foregroundStyle(Color.TS.textPrimary)
                .textSelection(.enabled)
                .fixedSize(horizontal: false, vertical: true)
                .frame(maxWidth: .infinity, alignment: .leading)
            UserImpersonateButton(model: model.rowModel(for: subject.subject))
        }
        .padding(.vertical, TSSpacing.md)
        .frame(maxWidth: .infinity, alignment: .leading)
        .accessibilityElement(children: .contain)
        .accessibilityIdentifier("users-page-row-\(subject.subject)")
    }
}

#if DEBUG
    #Preview("Subjects") {
        UsersPage(model: .sample())
            .teslaSyncTheme()
    }

    #Preview("Loading") {
        UsersPage(
            model: UsersPageModel(
                statusProvider: InMemoryImpersonationStatusProvider(
                    initial: .loaded(ImpersonationStatus(mode: .restricted))
                ),
                candidatesProvider: InMemoryUsersCandidatesProvider(initial: .loading)
            )
        )
        .teslaSyncTheme()
    }

    #Preview("Empty") {
        UsersPage(model: .sample(candidates: .loaded([])))
            .teslaSyncTheme()
    }

    #Preview("Error") {
        UsersPage(
            model: UsersPageModel(
                statusProvider: InMemoryImpersonationStatusProvider(
                    initial: .loaded(ImpersonationStatus(mode: .restricted))
                ),
                candidatesProvider: InMemoryUsersCandidatesProvider(initial: .failed(message: "Request timed out"))
            )
        )
        .teslaSyncTheme()
    }

    #Preview("Open mode") {
        UsersPage(
            model: UsersPageModel(
                statusProvider: InMemoryImpersonationStatusProvider(initial: .loaded(ImpersonationStatus(mode: .open))),
                candidatesProvider: InMemoryUsersCandidatesProvider(initial: .loaded([]))
            )
        )
        .teslaSyncTheme()
    }

    #Preview("Active (rows disabled)") {
        UsersPage(
            model: .sample(status: ImpersonationStatus(mode: .restricted, activeSubject: "ak-other-admin"))
        )
        .teslaSyncTheme()
    }
#endif

File: apps/android/app/src/main/kotlin/com/teslasync/admin/UsersPage.kt
// The native Jetpack Compose + Material 3 UsersPage admin surface — a parity port of
// web/src/features/admin/pages/UsersPage.tsx, the admin Subjects (impersonation targets) list. It reproduces the
// web page's single GlassPanel and every branch the web component renders inside it: the open-mode notice (web
// `data-testid="users-page-open-mode"`), the loading spinner, the hard-error retry, the no-subjects empty state,
// and the subjects list whose rows pair the opaque subject identifier with the shared UserImpersonateButton
// feature view (web `<UserImpersonateButton subject disabled={active} />`). Every visible string resolves from
// the generated res/values catalog (ADR-014); the opaque subject carries no display-unit fields, so there is no
// SI conversion (S5).
//
// Composition mirrors the sibling admin surfaces: [UsersPage] is the stateful entry (constructs the view-model
// over the host-wired source, records the one-shot `view.opened` diagnostic, collects the feeds);
// [UsersPageContent] is the stateless render layer that draws the title/subtitle header and the panel's
// open-mode / loading / error / empty / success surface. All derivation lives in the framework-free model
// (UsersPageModel.kt); this file only resolves i18n + draws.
//
// `InvalidPackageDeclaration` is suppressed: the mandated surface directory (com/teslasync/admin) diverges from
// the `io.teslasync.android.*` package the rest of the app uses.
@file:Suppress("InvalidPackageDeclaration")

package io.teslasync.android.admin.users

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.stringResource
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import io.teslasync.android.R
import io.teslasync.android.components.feedback.EmptyState
import io.teslasync.android.components.feedback.ErrorDisplay
import io.teslasync.android.components.feedback.Spinner
import io.teslasync.android.components.feedback.SpinnerSize
import io.teslasync.android.components.ui.BodyText
import io.teslasync.android.components.ui.CodeText
import io.teslasync.android.components.ui.GlassPanel
import io.teslasync.android.components.ui.PageTitle
import io.teslasync.android.components.ui.PanelPadding
import io.teslasync.android.data.LocalDataContainer
import io.teslasync.android.data.UiState
import io.teslasync.android.featureviews.userimpersonatebutton.UserImpersonateButton
import io.teslasync.android.ui.theme.generated.Spacing
import io.teslasync.shared.core.diagnostics.Logger
import io.teslasync.shared.core.presentation.impersonation.ImpersonationCandidate
import io.teslasync.shared.core.presentation.impersonation.ImpersonationCandidatesResponse

// ── Test tags (web `data-testid` parity) ──────────────────────────────────────────────────────────────────────
private const val TAG_OPEN_MODE = "users-page-open-mode"
private const val TAG_LOADING = "users-page-loading"
private const val TAG_ERROR = "users-page-error"
private const val TAG_EMPTY = "users-page-empty"
private const val TAG_LIST = "users-page-list"

private fun rowTag(subject: String): String = "users-page-row-$subject"

/** The page's interaction callbacks, wired to the [UsersPageViewModel] (web event handlers). */
data class UsersPageActions(
    val onStart: (subject: String) -> Unit,
    val onRetry: () -> Unit,
)

// ── Stateful entry points ───────────────────────────────────────────────────────────────────────────────────

/**
 * Stateful entry: constructs the [UsersPageViewModel] over the supplied [source] (the host wires the shared S8
 * [io.teslasync.shared.core.presentation.impersonation.ImpersonationStore] via [asUsersPageSource]). The
 * view-model is keyed by this surface's slug so it is scoped to the navigation entry. [logger] defaults to the
 * app's redacting logger.
 */
@Composable
fun UsersPage(
    source: UsersPageSource,
    modifier: Modifier = Modifier,
    logger: Logger = LocalDataContainer.current.logger,
) {
    val vm: UsersPageViewModel =
        viewModel(
            key = UsersPageRegistration.SLUG,
            factory = viewModelFactory { initializer { UsersPageViewModel(source, logger) } },
        )
    UsersPage(viewModel = vm, modifier = modifier)
}

/** Stateful entry: records the one-shot `view.opened` diagnostic and binds the feeds to the stateless content. */
@Composable
fun UsersPage(
    viewModel: UsersPageViewModel,
    modifier: Modifier = Modifier,
) {
    LaunchedEffect(viewModel) { viewModel.recordViewOpened() }

    val openMode by viewModel.isOpenMode.collectAsStateWithLifecycle()
    val active by viewModel.isActive.collectAsStateWithLifecycle()
    val candidatesState by viewModel.candidatesState.collectAsStateWithLifecycle()
    val startingSubject by viewModel.startingSubject.collectAsStateWithLifecycle()
    val actions =
        remember(viewModel) {
            UsersPageActions(onStart = viewModel::startImpersonation, onRetry = viewModel::retry)
        }

    UsersPageContent(
        openMode = openMode,
        candidatesState = candidatesState,
        active = active,
        startingSubject = startingSubject,
        actions = actions,
        modifier = modifier,
    )
}

// ── Stateless content ───────────────────────────────────────────────────────────────────────────────────────

/**
 * The stateless page body: the title/subtitle header (web `PageContainer` chrome) above the single GlassPanel
 * whose content branches exactly as the web component does — open-mode notice first (web `open`), then the
 * candidates feed's loading / error / empty / success surface. The panel keeps no inner padding so each branch
 * owns its own spacing, mirroring the web page's per-branch `p-6` / `p-8` / row `px-4 py-3`.
 */
@Composable
fun UsersPageContent(
    openMode: Boolean,
    candidatesState: UiState<ImpersonationCandidatesResponse>,
    active: Boolean,
    startingSubject: String?,
    actions: UsersPageActions,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier =
            modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(Spacing.lg),
        verticalArrangement = Arrangement.spacedBy(Spacing.lg),
    ) {
        UsersPageHeader()

        GlassPanel(padding = PanelPadding.None) {
            when {
                // GlassPanel1 — open-mode notice (web `open ? <div data-testid="users-page-open-mode">`).
                openMode -> OpenModeNotice()

                // loading (web `candidates.isLoading ? <Spinner />`).
                candidatesState.isLoading -> UsersLoadingState()

                // error (web `candidates.isError ? <ErrorDisplay onRetry={candidates.refetch} />`).
                candidatesState.isError -> UsersErrorState(onRetry = actions.onRetry)

                // empty (web `subjects.length === 0 ? <EmptyState />`).
                candidatesState.isEmpty -> UsersEmptyState()

                // success (web `<ul data-testid="users-page-list">` of subject rows).
                else ->
                    SubjectsList(
                        subjects = candidatesState.data?.subjects() ?: emptyList(),
                        active = active,
                        startingSubject = startingSubject,
                        onStart = actions.onStart,
                    )
            }
        }
    }
}

/** The page header — the title + muted subtitle (web `PageContainer` title/subtitle). */
@Composable
private fun UsersPageHeader() {
    Column(verticalArrangement = Arrangement.spacedBy(Spacing.xs)) {
        PageTitle(stringResource(R.string.translation_impersonation_users_title))
        BodyText(
            stringResource(R.string.translation_impersonation_users_subtitle),
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

// ── Panel branches ──────────────────────────────────────────────────────────────────────────────────────────

/**
 * Open-mode notice — the install runs without forward-auth, so per-user identity is unavailable (web
 * `impersonation.users.openMode`). Shown in place of the list, never as a blank region.
 */
@Composable
private fun OpenModeNotice() {
    Column(
        modifier =
            Modifier
                .fillMaxWidth()
                .testTag(TAG_OPEN_MODE)
                .padding(Spacing.lg),
    ) {
        BodyText(
            stringResource(R.string.translation_impersonation_users_openMode),
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

/** First-load surface — a centered spinner so the panel region is never blank (web `<Spinner />`). */
@Composable
private fun UsersLoadingState() {
    Column(
        modifier =
            Modifier
                .fillMaxWidth()
                .testTag(TAG_LOADING)
                .padding(Spacing.xl2),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Spinner(size = SpinnerSize.Md)
    }
}

/** Hard-error surface with a retry affordance (web `<ErrorDisplay error onRetry={candidates.refetch} />`). */
@Composable
private fun UsersErrorState(onRetry: () -> Unit) {
    ErrorDisplay(
        message = stringResource(R.string.translation_error_serverError_message),
        title = stringResource(R.string.translation_error_serverError_title),
        onRetry = onRetry,
        retryLabel = stringResource(R.string.translation_common_retry),
        modifier = Modifier.fillMaxWidth().testTag(TAG_ERROR),
    )
}

/** No-subjects surface — there is no other active subject to impersonate (web `<EmptyState />`). */
@Composable
private fun UsersEmptyState() {
    EmptyState(
        title = stringResource(R.string.translation_impersonation_users_emptyTitle),
        message = stringResource(R.string.translation_impersonation_users_emptyMessage),
        modifier = Modifier.fillMaxWidth().testTag(TAG_EMPTY),
    )
}

/**
 * The subjects list — the web `<ul data-testid="users-page-list">`: each row pairs the monospace subject
 * identifier with the shared UserImpersonateButton, separated by hairline dividers (web `divide-y`). The button
 * is disabled while a session is already active (web `disabled={active}`) and shows its in-flight state for the
 * row whose start is pending (web per-button `startMut.isPending`).
 */
@Composable
private fun SubjectsList(
    subjects: List<ImpersonationCandidate>,
    active: Boolean,
    startingSubject: String?,
    onStart: (subject: String) -> Unit,
) {
    Column(
        modifier =
            Modifier
                .fillMaxWidth()
                .testTag(TAG_LIST),
    ) {
        subjects.forEachIndexed { index, candidate ->
            if (index > 0) {
                HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)
            }
            SubjectRow(
                candidate = candidate,
                active = active,
                starting = startingSubject == candidate.subject,
                onStart = onStart,
            )
        }
    }
}

/** One subject row — the monospace identifier beside the per-row impersonate action (web `<li>`). */
@Composable
private fun SubjectRow(
    candidate: ImpersonationCandidate,
    active: Boolean,
    starting: Boolean,
    onStart: (subject: String) -> Unit,
) {
    Row(
        modifier =
            Modifier
                .fillMaxWidth()
                .testTag(rowTag(candidate.subject))
                .padding(horizontal = Spacing.md, vertical = Spacing.sm),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.md),
    ) {
        CodeText(candidate.subject, modifier = Modifier.weight(1f))
        UserImpersonateButton(
            subject = candidate.subject,
            starting = starting,
            onStart = onStart,
            disabled = active,
        )
    }
}
